modules/apigee_edge_teams/src/EventSubscriber/TeamStatusWarningSubscriber.php

```diff
@@ -100,9 +86,7 @@ class TeamStatusWarningSubscriber implements EventSubscriberInterface {
   public function __construct(AccountInterface $current_user, RouteMatchInterface $route_match, EntityTypeManagerInterface $entity_type_manager, TeamMembershipManagerInterface $team_membership_manager, MessengerInterface $messenger, TranslationInterface $string_translation) {
     $this->routeMatch = $route_match;
     $this->currentUser = $current_user;
-    $this->teamStorage = $entity_type_manager->getStorage('team');
```

Comment:
These are unrelated changes to the PR, please remove them.

Reply:
This is related as it prevents a clean uninstall of the `apigee_edge_apidocs` module. Accessing methods that accesses entity type discovery in a service constructor prevents clearing the entity type cache. See the following commit message: https://github.com/apigee/apigee-edge-drupal/pull/198/commits/b49093a592c1077dbe283367ad889dc230721f13

Reply:
If I remove @Jaesin 's changes I get the following error on uninstall, so these changes need to be in here:

```
The website encountered an unexpected error. Please try again later.</br></br><em class="placeholder">Drupal\Component\Plugin\Exception\PluginNotFoundException</em>: The &quot;apidoc&quot; entity type does not exist. in <em class="placeholder">Drupal\Core\Entity\EntityTypeManager-&gt;getDefinition()</em> (line <em class="placeholder">150</em> of <em class="placeholder">core/lib/Drupal/Core/Entity/EntityTypeManager.php</em>). <pre class="backtrace">Drupal\Core\Entity\EntityTypeManager-&gt;getHandler(&#039;apidoc&#039;, &#039;views_data&#039;) (Line: 157)
Drupal\Core\Entity\EntityManager-&gt;getHandler(&#039;apidoc&#039;, &#039;views_data&#039;) (Line: 177)
views_views_data()
call_user_func_array(&#039;views_views_data&#039;, Array) (Line: 392)
Drupal\Core\Extension\ModuleHandler-&gt;invoke(&#039;views&#039;, &#039;views_data&#039;) (Line: 243)
Drupal\views\ViewsData-&gt;getData() (Line: 160)
Drupal\views\ViewsData-&gt;get(&#039;apidoc&#039;) (Line: 91)
Drupal\views\Plugin\Derivative\ViewsEntityRow-&gt;getDerivativeDefinitions(Array) (Line: 101)
Drupal\Component\Plugin\Discovery\DerivativeDiscoveryDecorator-&gt;getDerivatives(Array) (Line: 87)
Drupal\Component\Plugin\Discovery\DerivativeDiscoveryDecorator-&gt;getDefinitions() (Line: 284)
Drupal\Core\Plugin\DefaultPluginManager-&gt;findDefinitions() (Line: 175)
Drupal\Core\Plugin\DefaultPluginManager-&gt;getDefinitions() (Line: 147)
views_theme(Array, &#039;module&#039;, &#039;views&#039;, &#039;core/modules/views&#039;) (Line: 447)
Drupal\Core\Theme\Registry-&gt;processExtension(Array, &#039;views&#039;, &#039;module&#039;, &#039;views&#039;, &#039;core/modules/views&#039;) (Line: 334)
Drupal\Core\Theme\Registry-&gt;build() (Line: 233)
Drupal\Core\Theme\Registry-&gt;get() (Line: 86)
Drupal\Core\Utility\ThemeRegistry-&gt;initializeRegistry() (Line: 67)
Drupal\Core\Utility\ThemeRegistry-&gt;__construct(&#039;theme_registry:runtime:seven&#039;, Object, Object, Array, 1) (Line: 253)
Drupal\Core\Theme\Registry-&gt;getRuntime() (Line: 142)
Drupal\Core\Theme\ThemeManager-&gt;render(Array, Array) (Line: 437)
Drupal\Core\Render\Renderer-&gt;doRender(Array, ) (Line: 195)
Drupal\Core\Render\Renderer-&gt;render(Array, ) (Line: 226)
Drupal\Core\Render\MainContent\HtmlRenderer-&gt;Drupal\Core\Render\MainContent\{closure}() (Line: 582)
Drupal\Core\Render\Renderer-&gt;executeInRenderContext(Object, Object) (Line: 227)
Drupal\Core\Render\MainContent\HtmlRenderer-&gt;prepare(Array, Object, Object) (Line: 117)
Drupal\Core\Render\MainContent\HtmlRenderer-&gt;renderResponse(Array, Object, Object) (Line: 90)
Drupal\Core\EventSubscriber\MainContentViewSubscriber-&gt;onViewRenderArray(Object, &#039;kernel.view&#039;, Object)
call_user_func(Array, Object, &#039;kernel.view&#039;, Object) (Line: 111)
Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher-&gt;dispatch(&#039;kernel.view&#039;, Object) (Line: 156)
Symfony\Component\HttpKernel\HttpKernel-&gt;handleRaw(Object, 1) (Line: 68)
Symfony\Component\HttpKernel\HttpKernel-&gt;handle(Object, 1, 1) (Line: 57)
Drupal\Core\StackMiddleware\Session-&gt;handle(Object, 1, 1) (Line: 47)
Drupal\Core\StackMiddleware\KernelPreHandle-&gt;handle(Object, 1, 1) (Line: 106)
Drupal\page_cache\StackMiddleware\PageCache-&gt;pass(Object, 1, 1) (Line: 85)
Drupal\page_cache\StackMiddleware\PageCache-&gt;handle(Object, 1, 1) (Line: 47)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware-&gt;handle(Object, 1, 1) (Line: 52)
Drupal\Core\StackMiddleware\NegotiationMiddleware-&gt;handle(Object, 1, 1) (Line: 23)
Stack\StackedHttpKernel-&gt;handle(Object, 1, 1) (Line: 693)
Drupal\Core\DrupalKernel-&gt;handle(Object) (Line: 19)
</pre>
```

Reply:
It would have been great if someone had validated #141 before this got merged.

Reply:
> A note about services that access the entity type manager in their constructors. Of course it's OK to inject the entity type manager in a service constructor but accessing any methods that result in `getDefinitions` being called should not be used in a service constructor. When a module is uninstalled, the last thing that happens is the service container is rebuilt. This happens when the module that is being uninstalled is still discoverable. Any entities the module defines will be cached when `getDefinitions` is called on the entity type manager.
> 
> This results in a `PluginNotFoundException` exception for subsequent page requests. The only way to get back to a working site is to clear the cache which would have to happen through drush. That is fine for developers but it is a problem for anyone that doesn't have drush or the correct alias set up.

@Jaesin Thanks for the investigation and the detailed answer but you haven't captured the _real_ root of the problem. Generally speaking, there should not be any problem with retrieving an instance of a class from an object repository (EntityTypeManager) in a constructor - unless there is a bug somewhere in the system. This is what I tried to convince you in #140 and #141 and this is the reason why I continued my investigation regarding this problem today. Based on my findings this Drupal core issue causes the problem: https://www.drupal.org/project/drupal/issues/3056604

When you uninstall the Apigee Edge Teams module then it gets uninstalled, you can see in the logs: `apigee_edge_teams module uninstalled.`. After that HTTP Kernel triggers the `KernelEvents::RESPONSE ` event and the :hankey: event dispatcher calls the uninstalled Apigee Edge Teams module's `TeamStatusWarningSubscriber` event subscriber, but it should not.

![screeshot-2019-05-23-11-08-39](https://user-images.githubusercontent.com/1755573/58249929-6ff26600-7d60-11e9-8847-d17a6c00db61.png)

(Stack trace captured after the Teams module uninstalled and the request has not been resolved. This confirms that the `TeamStatusWarningSubscriber` gets called because it is still registered in that instance of the event dispatcher and it fails because the "team" and "team_app" entity types have been unregistered already.)

Reply:
I found the same thing. Besides the `TeamStatusWarningSubscriber`, the `DeveloperStatusWarningSubscriber` causes the same issue. The teams module wasn't installed in the case of the error described in this issue.

This is not likely to get fixed in Drupal core any time soon (see: https://www.drupal.org/project/drupal/issues/3056604#comment-13118160). It is true that this results from issues in Drupal core but to be practical, we should work around it by not accessing methods that would cause the entity type cache to be rebuilt when the container is built. There aren't any **services** in Drupal core that use this pattern even through the entity type manager is used extensively in services so we should avoid it as well. Although the problem is showing itself while uninstalling modules, there could be edge cases where it shows up again.

Reply:
But entity type cache is not being rebuilt when you call `$entity_type_manager->getStorag()` in the event's constructor... it had been rebuilt when the module uninstalled (this is what [this method does](https://github.com/drupal/core/blob/8.7.2/lib/Drupal/Core/Extension/ModuleInstaller.php#L555)) and after that the only problem that Drupal core calls an uninstalled event dispatcher which tries to retrieve an uninstalled entity type. It does not matter where it tries to retrieve it, in its constructor or in the method that handles the event, it fails. (It does not even matter that this is an event subscriber, any code that would run after a module has been uninstalled and would try to retrieve an uninstalled entity type would fail.) This is what [the video](https://drive.google.com/open?id=1jQKtoMerpQXvuRQtQ_2KReJVl7u62lu1) above confirms. 

> This is not likely to get fixed in Drupal core any time soon

I'll work on this, just as I worked on the JSON API related bugs to ensure this gets fixed sooner as you would expect.

Let's continue this conversation in #203.